File: Swift/LeetCode/LeetCode/ImplementStrStr.swift
//
//  ImplementStrStr.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 25/11/2017.
//  Copyright © 2017 lizhaotailang. All rights reserved.
//
//  Implement strStr().
//
//  Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
//
//  Accepted. See [ImplementStrStrTests](./LeetCodeTest/ImplementStrStr.swift) for the test cases.
//

import Foundation

class ImplementStrStr {
    
    func strStr(_ haystack: String, _ needle: String) -> Int {
        if needle.count > haystack.count {
            return -1
        }
        if needle.count == 0 || haystack.count == 0 {
            return 0
        }
        
        var i = 0
        while i <= (haystack.count - needle.count) {
            if String(Array(haystack)[i...(i + needle.count - 1)]) == needle {
                return i
            }
            i += 1
        }
        
        return -1
    }
    
}

File: Golang/ImplementStrStr/ImplementStrStr.go
package ImplementStrStr

//Implement strStr().
//Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
//
//Accepted.

func strStr(haystack string, needle string) int {
	needleLength := len(needle)
	haystackLength := len(haystack)
	if needleLength > haystackLength {
		return -1
	}

	if haystackLength == 0 || needleLength == 0 {
		return 0
	}

	for i := 0; i <= haystackLength-needleLength; i++ {
		if haystack[i:i+needleLength] == (needle) {
			return i
		}
	}

	return -1
}
